File: src/main/java/io/github/colintimbarndt/chat_emotes/data/unicode/joiner/UnicodeJoiner.java
package io.github.colintimbarndt.chat_emotes.data.unicode.joiner;

import io.github.colintimbarndt.chat_emotes.data.unicode.pattern.MatchResult;
import io.github.colintimbarndt.chat_emotes.data.unicode.pattern.UnicodePattern;
import io.github.colintimbarndt.chat_emotes.data.unicode.pattern.PatternSlot;
import org.jetbrains.annotations.NotNull;

import java.text.ParseException;
import java.util.ArrayList;
import java.util.List;

public class UnicodeJoiner {
    private final ArrayList<JoinerSlot> slots = new ArrayList<>();
    private final ArrayList<String> literals = new ArrayList<>();

    private UnicodeJoiner() {}

    public static @NotNull UnicodeJoiner parse(@NotNull String s) throws ParseException {
        int i = 0;
        char c;
        final int sz = s.length();
        final var expr = new UnicodeJoiner();
        int currentLiteral = 0;
        for(; i < sz; i++) {
            c = s.charAt(i);
            if (c == '{') {
                if (i == 0) expr.literals.add("");
                else expr.literals.add(s.substring(currentLiteral, i));
                final int start = i;
                i = s.indexOf('}', start);
                if (i == -1) throw new ParseException("Unclosed Slot: " + s, start);
                final String content = s.substring(start + 1, i);
                expr.slots.add(JoinerSlot.parse(content, start));
                currentLiteral = i + 1;
            }
        }
        expr.literals.add(s.substring(currentLiteral, i));
        return expr;
    }

    public boolean isCompatibleWith(UnicodePattern pattern) {
        final var mss = pattern.getSlots().iterator();
        PatternSlot ms = null;
        int m = 0;
        for(var ts : slots) {
            int c = ts.count();
            if (m == 0) {
                if(!mss.hasNext()) return false;
                ms = mss.next();
                m = ms.range.min();
                if (!ts.acceptsWidth(ms.width())) return false;
            }
            if (ts.sameMatcher()) {
                if (ts.requiresName() && !ms.hasName()) return false;
                if (c == Integer.MAX_VALUE) {
                    // #& consumes the whole captures
                    m = 0;
                    continue;
                }
                if (ms.range.min() != c && ms.range.max() != c)
                    return false; // #X must consume whole captures
            } else {
                if (c == Integer.MAX_VALUE) {
                    // #* consumes all remaining captures
                    final boolean reqName = ts.requiresName();
                    while (mss.hasNext()) {
                        ms = mss.next();
                        if (!ts.acceptsWidth(ms.width())
                                || (reqName && !ms.hasName()))
                            return false;
                    }
                    return true;
                }
                if (ms.range.min() != ms.range.max())
                    return false; // Match must capture a consistent amount
                if (ts.requiresName() && !ms.hasName())
                    return false;
                m -= ts.count();
                while (m < 0) {
                    if (!mss.hasNext()) return false;
                    ms = mss.next();
                    if (ms.range.min() != ms.range.max())
                        return false;
                    if (ts.requiresName() && !ms.hasName())
                        return false;
                    m += ms.range.min();
                }
            }
        }
        return !mss.hasNext();
    }

    public String evaluate(final @NotNull List<MatchResult> @NotNull[] captures) {
        final StringBuilder builder = new StringBuilder();
        int sz = slots.size();
        int i = 0; // #slot
        int c = 0; // #capture
        int ci = 0; // #capture index
        builder.append(literals.get(i));
        if (captures.length == 0) return builder.toString();
        var capture = captures[c++];
        Outer: while (i < sz) {
            var slot = slots.get(i);
            boolean joiner = false;
            int consumed = 0;
            while (consumed < slot.count()) {
                if (ci == capture.size()) {
                    if (c == captures.length) break Outer;
                    ci = 0;
                    capture = captures[c++];
                    if (slot.sameMatcher()) break;
                }
                if (joiner) builder.append(slot.delimiter());
                joiner = true;
                slot.transform(capture.get(ci), builder);
                consumed++;
                ci++;
            }
            i++;
            builder.append(literals.get(i));
        }
        return builder.toString();
    }
}

File: src/main/kotlin/io/github/colintimbarndt/chat_emotes/data/unicode/joiner/UnicodeJoiner.kt
package io.github.colintimbarndt.chat_emotes.data.unicode.joiner

import io.github.colintimbarndt.chat_emotes.data.unicode.joiner.JoinerSlot.Companion.parse
import io.github.colintimbarndt.chat_emotes.data.unicode.pattern.MatchResult
import io.github.colintimbarndt.chat_emotes.data.unicode.pattern.PatternSlot
import io.github.colintimbarndt.chat_emotes.data.unicode.pattern.UnicodePattern
import java.text.ParseException

class UnicodeJoiner private constructor() {
    private val slots = ArrayList<JoinerSlot>()
    private val literals = ArrayList<String>()
    fun isCompatibleWith(pattern: UnicodePattern): Boolean {
        val patternSlots: Iterator<PatternSlot> = pattern.getCapturedSlots().iterator()
        var patternSlot: PatternSlot? = null
        var minRem = 0
        var maxOffset = 0
        if (patternSlots.hasNext()) {
            patternSlot = patternSlots.next()
            minRem = patternSlot.range.min
            maxOffset = patternSlot.range.max - minRem
        } else {
            if (slots.isNotEmpty()) return false
        }
        for (ts in slots) {
            val c = ts.count
            if (minRem == 0) {
                if (maxOffset != 0) return false
                if (!patternSlots.hasNext()) return false
                patternSlot = patternSlots.next()
                minRem = patternSlot.range.min
                maxOffset = patternSlot.range.max - minRem
            }
            patternSlot!!
            if (!ts.acceptsWidth(patternSlot.width)) return false
            if (ts.sameMatcher) {
                if (ts.requiresName && !patternSlot.hasName) return false
                if (c == Int.MAX_VALUE) {
                    // #& consumes the whole current captures
                    minRem = 0
                    maxOffset = 0
                    continue
                }
                if (minRem < c) return false
                minRem -= c
            } else {
                // whole-matcher
                if (c == Int.MAX_VALUE) {
                    // #* consumes all remaining captures
                    val reqName = ts.requiresName
                    while (patternSlots.hasNext()) {
                        patternSlot = patternSlots.next()
                        if (!ts.acceptsWidth(patternSlot.width)) return false
                        if (reqName && !patternSlot.hasName) return false
                    }
                    return true
                }
                if (maxOffset != 0) return false // Match must capture a consistent amount
                if (ts.requiresName && !patternSlot.hasName) return false
                minRem -= ts.count
                while (minRem < 0) {
                    if (!patternSlots.hasNext()) return false
                    patternSlot = patternSlots.next()
                    if (!ts.acceptsWidth(patternSlot.width)) return false
                    if (patternSlot.range.min != patternSlot.range.max) return false
                    if (ts.requiresName && !patternSlot.hasName) return false
                    minRem += patternSlot.range.min
                }
            }
        }
        return !patternSlots.hasNext() && minRem == 0
    }

    fun evaluate(captures: Array<List<MatchResult>>): String {
        val builder = StringBuilder()
        val sz = slots.size
        var i = 0 // #slot
        var c = 0 // #capture
        var ci = 0 // #capture index
        builder.append(literals[i])
        if (captures.isEmpty()) return builder.toString()
        var capture = captures[c++]
        Outer@ while (i < sz) {
            val slot = slots[i]
            var joiner = false
            var consumed = 0
            while (consumed < slot.count) {
                if (ci == capture.size) {
                    if (c == captures.size) break@Outer
                    ci = 0
                    capture = captures[c++]
                    if (slot.sameMatcher) break
                }
                if (joiner) builder.append(slot.delimiter)
                joiner = true
                slot.transform(capture[ci], builder)
                consumed++
                ci++
            }
            i++
            builder.append(literals[i])
        }
        return builder.toString()
    }

    companion object {
        @Throws(ParseException::class)
        fun parse(s: String): UnicodeJoiner {
            var i = 0
            var c: Char
            val sz = s.length
            val expr = UnicodeJoiner()
            var currentLiteral = 0
            while (i < sz) {
                c = s[i]
                if (c == '{') {
                    if (i == 0) expr.literals.add("") else expr.literals.add(s.substring(currentLiteral, i))
                    val start = i
                    i = s.indexOf('}', start)
                    if (i == -1) throw ParseException("Unclosed Slot: $s", start)
                    val content = s.substring(start + 1, i)
                    expr.slots.add(parse(content, start))
                    currentLiteral = i + 1
                }
                i++
            }
            expr.literals.add(s.substring(currentLiteral, i))
            return expr
        }
    }
}